volumes/miovision/sql/function-find_gaps.sql

```diff
@@ -1,116 +1,199 @@
+            - datetime_bin > interval '1 minute' AS bin_break, --True means gap between bins is larger than 1 minute
+        --weekend is needed to determine which gapsize_lookup to use. 
+        CASE
+            WHEN date_part('isodow', datetime_bin) <= 5
+                AND hol.holiday IS NULL THEN False
+            ELSE True
+        END as weekend
+    FROM fluffed_data
+    LEFT JOIN ref.holiday AS hol ON hol.dt = fluffed_data.datetime_bin::date
+    GROUP BY
+        intersection_uid, 
+        datetime_bin,
+        weekend
+),
+
+gaps AS (
+	--find gaps longer than the threshold from gapsize_lookup and cross join with
+    --generate_series output to enable easily joining with 15 minute datetime_bins.
+	INSERT INTO miovision_api.unacceptable_gaps(
+		intersection_uid, gap_start, gap_end, gap_minutes_total, allowable_total_gap_threshold, 
+        datetime_bin, gap_minutes_15min, avg_historical_total_vol, avg_historical_veh_vol
+	)    
+    SELECT
+		bt.intersection_uid,
+		bt.gap_start,
+		bt.gap_end,
+		gm.gap_minutes_total,
+        gl.gap_tolerance AS allowable_total_gap_threshold,
+        bins.datetime_bin,		
+        round(gm.gap_minutes_15min) AS gap_minutes_15min,
+        --avg historical volumes during the gaps to use for imputation
+        --or decision on summary data quality
+        round(gl.avg_vol * gm.gap_minutes_15min / 60) AS avg_historical_total_vol,
+        round(gl.avg_vehicle_vol * gm.gap_minutes_15min / 60) AS avg_historical_veh_vol
+	FROM bin_times AS bt
+    --match gaps to the 15 minute bins they intersect
+    LEFT JOIN generate_series(
+            --catch gaps overlapping days
+            start_date::timestamp - interval '15 minutes',
+            start_date::timestamp + interval '1 day',
+            interval '15 minutes'
+    ) AS bins(datetime_bin) ON
+        bins.datetime_bin >= datetime_bin(bt.gap_start, 15)
+        AND bins.datetime_bin < datetime_bin_ceil(bt.gap_end, 15)
+    --find the acceptable gap size based on historical lookback.
+	LEFT JOIN gapsize_lookup AS gl ON
+        gl.intersection_uid = bt.intersection_uid
+        AND gl.time_bin = date_part('hour', bins.datetime_bin)
+        AND gl.weekend = bt.weekend,
+	LATERAL (
+		SELECT
+            date_part('epoch', gap_end - gap_start) / 60::integer AS gap_minutes_total,
+            --find the gap minutes which occured only during the 15 minute bin (bins.datetime_bin)
+            EXTRACT (EPOCH FROM
+                 LEAST(bt.gap_end, bins.datetime_bin + interval '15 minutes')
+                 - GREATEST(bt.gap_start, bins.datetime_bin)
+                ) / 60 AS gap_minutes_15min
+	) AS gm
+	WHERE
+		gm.gap_minutes_total >= gl.gap_tolerance
+		AND bt.bin_break = True
+		AND bt.gap_end IS NOT NULL
+        --exclude gaps that are entirely without of todays interval
+        AND NOT(bins.datetime_bin + gm.gap_minutes_15min * '1 minute'::interval) <= start_date 
```

Comment:
why not use `gap_end` here?

Reply:
Ignore my previous comment. Don't you want to exclude `datetime_bin < start_date`?

Reply:
I had to think about this for a while. 
I think NOT(gap_end <= start_date) is actually correct. If it's a gap that overlaps the current day at all, we want to include the 23:45-00:00 datetime_bin as a gap, because it wouldn't be caught if only looking at yesterdays data.